core/src/main/java/com/sequenceiq/cloudbreak/service/freeipa/FreeIpaConfigProvider.java

```diff
@@ -44,13 +45,23 @@
     private Optional<InstanceMetaDataResponse> selectInstance(DescribeFreeIpaResponse freeIpaResponse) {
         List<InstanceMetaDataResponse> metaDataResponses = freeIpaResponse.getInstanceGroups().stream()
                 .flatMap(ig -> ig.getMetaData().stream()).collect(Collectors.toList());
+        Optional<InstanceMetaDataResponse> pgwInstance = selectPgwInstance(metaDataResponses);
         Optional<InstanceMetaDataResponse> instanceMetaDataResponse =
-                chooseFirstOrderedByFqdnAndByStateIfPresent(metaDataResponses, Optional.of(InstanceStatus.CREATED))
-                        .or(() -> chooseFirstOrderedByFqdnAndByStateIfPresent(metaDataResponses, Optional.empty()));
+                pgwInstance.or(() -> chooseFirstOrderedByFqdnAndByStateIfPresent(metaDataResponses, Optional.of(InstanceStatus.CREATED))
+                        .or(() -> chooseFirstOrderedByFqdnAndByStateIfPresent(metaDataResponses, Optional.empty())));
         LOGGER.info("Chosen instance for default FreeIPA server: {}", instanceMetaDataResponse);
         return instanceMetaDataResponse;
     }
 
+    private Optional<InstanceMetaDataResponse> selectPgwInstance(List<InstanceMetaDataResponse> metaDataResponses) {
+        List<InstanceMetaDataResponse> pgwInstances = metaDataResponses.stream().
+                filter(im -> InstanceMetadataType.GATEWAY_PRIMARY.equals(im.getInstanceType()))
```

Comment:
Question: isn't there only one instance with GATEWAY_PRIMARY imd type?

Reply:
should be, yes, but this way it's more defensive